File: clients/typescript/src/sync.ts
// Watermark-driven incremental mirror. Cold start streams a filtered snapshot;
// every run after that pulls only the change-feed delta since the last
// watermark. Applies post-images straight from the feed (no follow-up reads),
// tombstones removed keys, and advances the watermark only after the sink
// commits — so a crash mid-run re-processes idempotently rather than skipping.

import { PumperClient, type PumperClientConfig } from "./client.js";
import type {
  DatasetRef,
  MapContext,
  SyncProgress,
  SyncResult,
  SyncSink,
  WatermarkStore,
} from "./types.js";

export interface PumperSyncConfig<TRaw = unknown, TOut = TRaw> extends PumperClientConfig {
  /** The canonical dataset to mirror — prefer a unified dataset (e.g.
   *  `{ app: "grants", name: "unified" }`) so the product never re-normalizes. */
  dataset: DatasetRef;
  /** Repeatable `<path>:<op>:<value>` predicates, ANDed and pushed into SQL, so
   *  a product mirrors only its slice (e.g. `["$.status:eq:open"]`). */
  filter?: string[];
  watermark: WatermarkStore;
  sink: SyncSink<TOut>;
  /** Optional per-record transform on the way into the sink. Return null to drop
   *  a record. Omit for a straight canonical mirror. */
  map?: (raw: TRaw, ctx: MapContext) => TOut | null;
  /** Flush the sink every N records (snapshot mode). Default 500. */
  batchSize?: number;
  onProgress?: (p: SyncProgress) => void;
  /** Stop a long snapshot stream. */
  signal?: AbortSignal;
}

/** Lexicographic max of two RFC3339 watermarks — valid because Pumper stamps
 *  fixed-width UTC micros, so string order == chronological order. */
function laterIso(a: string | null, b: string): string {
  return a === null || b > a ? b : a;
}

class PumperSync<TRaw, TOut> {
  private readonly client: PumperClient;
  private readonly batchSize: number;

  constructor(private readonly cfg: PumperSyncConfig<TRaw, TOut>) {
    this.client = new PumperClient(cfg);
    this.batchSize = cfg.batchSize ?? 500;
  }

  /** Run one sync cycle: snapshot on cold start (no watermark), else incremental. */
  async run(): Promise<SyncResult> {
    const since = await this.cfg.watermark.get(this.cfg.dataset);
    return since === null ? this.snapshot() : this.incremental(since);
  }

  private toOut(raw: TRaw, ctx: MapContext): TOut | null {
    return this.cfg.map ? this.cfg.map(raw, ctx) : (raw as unknown as TOut);
  }

  private async snapshot(): Promise<SyncResult> {
    const { dataset, filter, sink, onProgress } = this.cfg;
    let upserts: Array<{ key: string; data: TOut }> = [];
    let tombstones: string[] = [];
    let upserted = 0;
    let tombstoned = 0;
    let watermark: string | null = null;

    const flush = async () => {
      if (upserts.length) {
        upserted += await sink.upsert(upserts);
        upserts = [];
      }
      if (tombstones.length) {
        tombstoned += await sink.tombstone(tombstones);
        tombstones = [];
      }
      onProgress?.({ mode: "snapshot", upserted, tombstoned });
    };

    for await (const rec of this.client.exportRecords<TRaw>(dataset, filter, this.cfg.signal)) {
      watermark = laterIso(watermark, rec.updated_at);
      if (rec.removed_at) {
        tombstones.push(rec.key);
      } else {
        const out = this.toOut(rec.data, { key: rec.key, updatedAt: rec.updated_at, change: "changed" });
        if (out !== null) upserts.push({ key: rec.key, data: out });
      }
      if (upserts.length >= this.batchSize || tombstones.length >= this.batchSize) await flush();
    }
    await flush();

    if (watermark !== null) await this.cfg.watermark.set(dataset, watermark);
    return { mode: "snapshot", upserted, tombstoned, watermark };
  }

  private async incremental(since: string): Promise<SyncResult> {
    const { dataset, sink } = this.cfg;
    // The feed is newest-first, so the FIRST revision we see for a key is its
    // latest — track seen keys and skip older duplicates so the latest wins.
    const seen = new Set<string>();
    const upserts: Array<{ key: string; data: TOut }> = [];
    const tombstones: string[] = [];
    let watermark = since;
    let cursor = "";

    for (;;) {
      const page = await this.client.changesPage<TRaw>(dataset, since, cursor);
      for (const rev of page.items) {
        watermark = laterIso(watermark, rev.created_at);
        if (seen.has(rev.key)) continue;
        seen.add(rev.key);
        if (rev.change === "removed") {
          tombstones.push(rev.key);
        } else if (rev.data !== null) {
          const out = this.toOut(rev.data, { key: rev.key, updatedAt: rev.created_at, change: rev.change });
          if (out !== null) upserts.push({ key: rev.key, data: out });
        }
      }
      if (!page.next_cursor) break;
      cursor = page.next_cursor;
    }

    const upserted = upserts.length ? await sink.upsert(upserts) : 0;
    const tombstoned = tombstones.length ? await sink.tombstone(tombstones) : 0;
    this.cfg.onProgress?.({ mode: "incremental", upserted, tombstoned });

    await this.cfg.watermark.set(dataset, watermark);
    return { mode: "incremental", upserted, tombstoned, watermark };
  }
}

/** Build a mirror for one canonical dataset. Call `.run()` on a schedule (cron,
 *  a trigger, `POST /api/cron/...`); each call advances the watermark. */
export function createPumperSync<TRaw = unknown, TOut = TRaw>(
  cfg: PumperSyncConfig<TRaw, TOut>,
): { run: () => Promise<SyncResult> } {
  const sync = new PumperSync(cfg);
  return { run: () => sync.run() };
}

File: clients/python/pumper_sync/sync.py
"""Watermark-driven incremental mirror — the Python twin of `sync.ts`.

Cold start streams a filtered snapshot; every run after that pulls only the
change-feed delta since the last watermark. Post-images are applied straight
from the feed (no follow-up record reads), removed keys are tombstoned, and the
watermark advances only AFTER the sink commits — so a crash mid-run
re-processes idempotently rather than skipping.

The loop is deliberately identical to the TypeScript one, step for step,
including the newest-first de-duplication. Two SDKs that mirror the same feed
with subtly different loops are two different products, and the difference
would only ever show up as a divergent mirror in somebody's database.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable, Literal, Protocol

from .client import PumperClient
from .generated import Json

SyncMode = Literal["snapshot", "incremental"]


@dataclass(frozen=True)
class DatasetRef:
    """A dataset address: ``<app>/<name>`` (e.g. ``grants/unified``)."""

    app: str
    name: str


@dataclass
class SyncResult:
    mode: SyncMode
    upserted: int
    tombstoned: int
    #: The watermark persisted at the end of this run. ``None`` when the dataset
    #: was empty and nothing advanced it — an honest "we learned nothing", not a
    #: zero.
    watermark: str | None


class WatermarkStore(Protocol):
    """Where the product persists the per-dataset sync watermark.

    The product owns storage (a row, a KV entry, a file); the SDK only reads and
    advances it. The value is an opaque RFC 3339 token — do not parse it.
    """

    def get(self, dataset: DatasetRef) -> str | None: ...

    def set(self, dataset: DatasetRef, watermark: str) -> None: ...


class SyncSink(Protocol):
    """The product's persistence boundary. Returns the count actually written,
    purely for reporting."""

    def upsert(self, records: list[tuple[str, Json]]) -> int: ...

    def tombstone(self, keys: list[str]) -> int: ...


@dataclass
class MemoryWatermark:
    """In-process watermark. For tests and one-shot scripts — a mirror that must
    survive a restart needs a real store, and using this one by accident would
    silently re-snapshot on every run."""

    values: dict[tuple[str, str], str] = field(default_factory=dict)

    def get(self, dataset: DatasetRef) -> str | None:
        return self.values.get((dataset.app, dataset.name))

    def set(self, dataset: DatasetRef, watermark: str) -> None:
        self.values[(dataset.app, dataset.name)] = watermark


def later_iso(a: str | None, b: str) -> str:
    """Lexicographic max of two RFC 3339 watermarks.

    Valid because Pumper stamps fixed-width UTC micros, so string order IS
    chronological order — no parsing, and therefore no timezone to get wrong.
    """
    return b if a is None or b > a else a


class PumperSync:
    """Mirror one canonical dataset. Call :meth:`run` on a schedule; each call
    advances the watermark."""

    def __init__(
        self,
        dataset: DatasetRef,
        watermark: WatermarkStore,
        sink: SyncSink,
        *,
        client: PumperClient | None = None,
        filters: list[str] | None = None,
        map_record: Callable[[Json, str], Json | None] | None = None,
        batch_size: int = 500,
    ) -> None:
        self.dataset = dataset
        self.watermark = watermark
        self.sink = sink
        self.client = client or PumperClient()
        self.filters = filters or []
        self.map_record = map_record
        self.batch_size = batch_size

    def run(self) -> SyncResult:
        since = self.watermark.get(self.dataset)
        return self._snapshot() if since is None else self._incremental(since)

    def _to_out(self, raw: Json, key: str) -> Json | None:
        return self.map_record(raw, key) if self.map_record else raw

    def _snapshot(self) -> SyncResult:
        upserts: list[tuple[str, Json]] = []
        tombstones: list[str] = []
        upserted = tombstoned = 0
        watermark: str | None = None

        def flush() -> None:
            nonlocal upserts, tombstones, upserted, tombstoned
            if upserts:
                upserted += self.sink.upsert(upserts)
                upserts = []
            if tombstones:
                tombstoned += self.sink.tombstone(tombstones)
                tombstones = []

        for rec in self.client.export_records(self.dataset.app, self.dataset.name, self.filters):
            watermark = later_iso(watermark, rec["updated_at"])
            if rec.get("removed_at"):
                tombstones.append(rec["key"])
            else:
                out = self._to_out(rec["data"], rec["key"])
                if out is not None:
                    upserts.append((rec["key"], out))
            if len(upserts) >= self.batch_size or len(tombstones) >= self.batch_size:
                flush()
        flush()

        if watermark is not None:
            self.watermark.set(self.dataset, watermark)
        return SyncResult("snapshot", upserted, tombstoned, watermark)

    def _incremental(self, since: str) -> SyncResult:
        # The feed is newest-first, so the FIRST revision seen for a key is its
        # latest — track seen keys and skip older duplicates so the latest wins.
        seen: set[str] = set()
        upserts: list[tuple[str, Json]] = []
        tombstones: list[str] = []
        watermark = since
        cursor = ""

        while True:
            page = self.client.changes_page(self.dataset.app, self.dataset.name, since, cursor)
            for rev in page.get("items", []):
                watermark = later_iso(watermark, rev["created_at"])
                if rev["key"] in seen:
                    continue
                seen.add(rev["key"])
                if rev["change"] == "removed":
                    tombstones.append(rev["key"])
                elif rev.get("data") is not None:
                    out = self._to_out(rev["data"], rev["key"])
                    if out is not None:
                        upserts.append((rev["key"], out))
            next_cursor = page.get("next_cursor")
            if not next_cursor:
                break
            cursor = next_cursor

        upserted = self.sink.upsert(upserts) if upserts else 0
        tombstoned = self.sink.tombstone(tombstones) if tombstones else 0
        self.watermark.set(self.dataset, watermark)
        return SyncResult("incremental", upserted, tombstoned, watermark)
